source/common/network/connection_impl.cc

```diff
@@ -269,7 +269,9 @@ void ConnectionImpl::onRead(uint64_t read_buffer_size) {
     read_end_stream_raised_ = true;
   }
 
-  filter_manager_.onRead();
+  if (!inDelayedClose()) {
```

Comment:
Yeah, I think I'll put an `ASSERT` there to make sure we definitely aren't tripping on this again and fix it via read disable.

Reply:
Agree that this is the fix, and I think I've seen some debug build ASSERTs in production that are due to this issue, so great to see it fixed.

@AndresGuedez IMO the way we handle `read_enabled_` inside `close()` is pretty fragile with multiple return points. Is there any way to simplify that code to make it less fragile and also handle this case?

Reply:
> @AndresGuedez IMO the way we handle `read_enabled_` inside `close()` is pretty fragile with multiple return points. Is there any way to simplify that code to make it less fragile and also handle this case?

Agreed. I'll file an issue for tracking and will work on a PR to refactor and minimize the opportunity for this type of issue to reoccur.

Reply:
Would it be reasonable to just unconditionally set `read_enabled_ = false` in `close()` for all paths? Is there ever a situation we want to keep reading.

Reply:
> Would it be reasonable to just unconditionally set read_enabled_ = false in close() for all paths? Is there ever a situation we want to keep reading.

This would be my initial inclination, but I would need to read the code in detail again to understand why it is the way it is if there is a reason. @AndresGuedez has been here more recently so might be able to quickly advise.

Reply:
Yeah, tests to follow once we're agreed on a solution. I agree that the fewer state variables we need to condition on the better. Happy to just switch to checking close state, but will waitfor @AndresGuedez to advise, this code is quite tricky and there are gotchas.